### ISSUE TITLE
No equivalent to escape_dn_chars to escape dn strings.

### ISSUE THREAD

**Issue by CnlPepper** [cannatag/ldap3#502]
Until recently I've been using the python-ldap package but have just started moving to ldap3 as it has a cleaner API and fewer dependencies. There is one function I can't seem to find. To prevent injections into dn strings when e.g. using bind and taking in a username, the username needs to be escaped to prevent users injecting data into the dn string e.g. if they supply "blah,abc=123" it would result in a modified dn string.

In python-ldap there is the escape_dn_chars function (see https://github.com/python-ldap/python-ldap/blob/master/Lib/ldap/dn.py), I feel this should be available in the ldap3.utils.dn module. I've found the equivalent filter escape function, but it does not cover all the characters that must be handled for a dn according to RFC 4514.

For now I've reimplemented the function locally, could this be added to the package if no equivalent exists?

**Comment by cannatag** [comment]
Hi, can you try the ``safe_dn()`` function in ldap3.utils.dn? It do some sanity check for DNs including escaping of illegal characters.

**Comment by CnlPepper** [comment]
I've tried safe_dn but it does not prevent values from containing commas etc... (see [10] below).
```
In [10]: sdn(['uid={}'.format('baduser,bad=stuff'), 'ou=foo', 'dc=bar'])
Out[10]: 'uid=baduser,bad=stuff,ou=foo,dc=bar'
```
I've reimplemented the code in python-ldap in my own code to ensure data I receive from users is escaped:

```
def escape_dn_chars(s):
    """
    Escape dn characters to prevent injection according to RFC 4514.
    """

    # '/' must be handled first or the escape slashes will be escaped!
    for char in ['\\', ',', '+', '"', '<', '>', ';', '=', '\x00']:
        s = s.replace(char, '\\' + char)

    if s[0] == '#' or s[0] == ' ':
        s = ''.join(('\\', s))

    if s[-1] == ' ':
        s = ''.join((s[:-1], '\\ '))

    return s
```

Example usage:
```
In [12]: 'uid={},ou=foo,dc=bar'.format(escape_dn_chars('baduser,bad=stuff'))
Out[12]: 'uid=baduser\\,bad\\=stuff,ou=foo,dc=bar'
```

On a browse around the web for how other people have solved this with ldap3, I've found a lot of codebases are incorrectly using the filter escape function which does not filter commas etc...

I'm wondering if it might be worth adding the escape dn function and update the examples in the documentation to demonstrate it's use when receiving data from untrusted sources to reduce the risk of people incorrectly inventing their own solutions.

**Comment by cannatag** [comment]
Hi, you're right. 
I've included your function in the ldap3.utils.dn module but changed the name to escape_rdn_chars because you must use it only on one part of the dn not on the whole dn itself:

```
>>> from ldap3.utils.dn import escape_rdn_chars
>>> a = 'baduser,bad=stuff'
>>> print(escape_rdn_chars(a))  # correct
'baduser\\,bad\\=stuff'
>>> b = 'cn='+a+',ou=users,o=test'
>>> print(escape_rdn_chars(b))  # wrong
>>> 'cn\\=baduser\\,bad\\=stuff\\,ou\\=utenti\\,o\\=test'
```

Do you think it's a good solution?
Bye,
Giovanni

**Comment by cannatag** [comment]
I've renamed ``escape_rdn_chars()`` to ``escape_rdn``. It looks simpler to me. I've also added a sidebar in the Connection page on ldap3.readthedocs.io.

**Comment by CnlPepper** [comment]
Hi Giovanni,

the solution looks good to me, many thanks for the quick response!

Best regards,
Alex

**Comment by cannatag** [comment]
Fixed in 2.5.

**Comment by sijanec** [comment]
What about newline and other control characters? Is it okay that they are left in the DN? I'm building a DN using `'uid=' + ldap3.utils.dn.escape_rdn(user_provided_string) + ',ou=users,dc=example'`, and the user provided string is obtained from the login form via HTTP, so it can contain a newline. Is this a problem?
